lib/rules/alias-model-in-controller.js

```diff
@@ -29,7 +29,9 @@ module.exports = {
     };
 
     return {
-      CallExpression(node) {
+      'CallExpression > MemberExpression[property.name="extend"]'(memberExpressionNode) {
```

Comment:
We should not push this complexity (checking for extends and parents and such) into each rule. Each rule should be able to make one simple util function call to check if the node is an Ember controller/component/etc, like this:

```
CallExpression(node) {
  // classic class
  if (ember.isEmberController(context, node)) {
    ...
  }
},

ClassDeclaration(node) {
  // native class
  if (ember.isEmberController(context, node)) {
    ...
  }
},
```

Reply:
I'm not sure I agree -- it seems like we have a specific `CallExpression` pattern we're looking for, which looks like

```
Foo.extend()
```

We have the ability to tell ESLint those details -- why would we want to check _every_ call expression if we don't have to?

Reply:
To my knowledge, there's no actual difference between using the more-specific [selector](https://eslint.org/docs/developer-guide/selectors) vs. doing it ourselves: either eslint does the checks or we do the checks. In general, I would agree with you that it's nice to take advantage of using more-specific selectors, but I don't think that's appropriate here, because:

1. Using both the selector matching and util function unnecessarily splits the logic for detecting components/controllers/etc into two separate places instead of consolidating the logic fully inside a single util function (and maybe this util function could even come from a third-party library someday).
2. Using the selector matching duplicates part of the detection logic across countless rules, which increases the chance for typos and inconsistencies between rules. 
3. Encapsulating all of the logic in a util function makes it easy to unit test it in its entirety.
4. The rule code is simpler when the detection details are delegated to the util function. We should optimize for simplicity and readability, and make it as easy to write rules as possible (which includes providing extremely simple util function APIs).